### ISSUE TITLE
[JENKINS-43894] Environment variables not resolved in Pipeline SCM -> Advanced clone behaviours -> Path of the reference repo

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3093]
<p>Environment variables are not resolved in 'Advanced clone behaviours' -> 'Path of the reference repo to use during clone' when configuring git reference repository in the Pipeline SCM section.</p>

<p>Steps to reproduce:</p>
<ul>
	<li>Create pipeline project</li>
	<li>In configure: fill in repository URL, Branch Specifier and Script Path (Jenkinsfile).</li>
	<li>In 'Additional Behaviours' select 'Advanced Clone Behaviours'. Set 'Path of the reference repo to use during clone' to some value depending on an environment variable, e.g. ${HOME}/jenkins/reference/big-git-repo.git</li>
	<li>At URL, create a git repository with Jenkinsfile. Put 'node('master') { checkout scm }' in the Jenkinsfile.</li>
	<li>Run the job. It will report: 

<pre>
Cloning repository ...URL...
 > git init ... # timeout=10
ERROR: Reference path does not exist: ${HOME}/jenkins/reference/big-git-repo.git</pre></li>
</ul>

<p>Would be very handy to have this feature. In my case, I want to use the 'checkout scm' shortcut in a pipeline script that runs several builds in parallel on different platforms. So I want to specify where a reference repository exists on a specific node. For instance, on a Windows node I want to put it in D:\jenkins\reference, on a Mac – to /Users/jenkins/reference and, finally, on linux – /home/jenkins/reference. I could set an environment variable ${REFERENCE} to the above values in node configuration and than use it in 'Path of the reference repo to use during clone' as ${REFERENCE}.</p>

<p> </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=atikhono">atikhono</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894" target="_blank">Environment variables not resolved in Pipeline SCM -> Advanced clone behaviours -> Path of the reference repo</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=atikhono">atikhono</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: git-plugin
<li><b>label(s)</b>: pipeline
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-10-24T12:41:10+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 11
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Environment variables are not resolved in 'Advanced clone behaviours' -&gt; 'Path of the reference repo to use during clone' when configuring git reference repository in the Pipeline SCM section.</p>

<p>Steps to reproduce:</p>
<ul>
	<li>Create pipeline project</li>
	<li>In configure: fill in repository URL, Branch Specifier and Script Path (Jenkinsfile).</li>
	<li>In 'Additional Behaviours' select 'Advanced Clone Behaviours'. Set 'Path of the reference repo to use during clone' to some value depending on an environment variable, e.g. ${HOME}/jenkins/reference/big-git-repo.git</li>
	<li>At URL, create a git repository with Jenkinsfile. Put 'node('master') { checkout scm }' in the Jenkinsfile.</li>
	<li>Run the job. It will report: 
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Cloning repository ...URL...
 &gt; git init ... # timeout=10
ERROR: Reference path does not exist: ${HOME}/jenkins/reference/big-git-repo.git</pre>
</div></div></li>
</ul>

<p>Would be very handy to have this feature. In my case, I want to use the 'checkout scm' shortcut in a pipeline script that runs several builds in parallel on different platforms. So I want to specify where a reference repository exists on a specific node. For instance, on a Windows node I want to put it in D:\jenkins\reference, on a Mac – to /Users/jenkins/reference and, finally, on linux – /home/jenkins/reference. I could set an environment variable ${REFERENCE} to the above values in node configuration and than use it in 'Path of the reference repo to use during clone' as ${REFERENCE}.</p>

<p> </p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins 2.89.3, workflow-job 2.18-SNAPSHOT</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
- _Remote links associated with this issue:_


  - [PR 502](https://github.com/jenkinsci/git-plugin/pull/502)
  - [PR 575](https://github.com/jenkinsci/git-plugin/pull/575)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=297399&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297399">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Originally, I asked about this behaviour here: <a href="https://issues.jenkins-ci.org/browse/JENKINS-28447" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-28447</a></p>

<p>Was told to submit an RFE for git plugin.</p>
</pre>
</details>

</ul>
</details>

<p>Originally, I asked about this behaviour here: <a href="https://issues.jenkins-ci.org/browse/JENKINS-28447" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-28447</a></p>

<p>Was told to submit an RFE for git plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=297488&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297488">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>As another alternate, since you're using pipeline, the git plugin could be extended to automatically use a reference repository.  The mercurial plugin uses that technique to reduce duplication on disc.</p>
</pre>
</details>

</ul>
</details>

<p>As another alternate, since you're using pipeline, the git plugin could be extended to automatically use a reference repository.  The mercurial plugin uses that technique to reduce duplication on disc.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=297558&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297558">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> Could you please clarify a bit how it works in the Mercurial plugin? I've no experience with Mercurial and can't find anything resembling reference repositories implementation in the source code.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> Could you please clarify a bit how it works in the Mercurial plugin? I've no experience with Mercurial and can't find anything resembling reference repositories implementation in the source code.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=297566&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297566">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I believe Jesse Glick described that the SCMFileSystem implementation in the Mercurial plugin uses the initial checkout of the repository (for polling purposes) as a cache for other copies on the disc.  The concept was what was interesting to me, more than the specifics of the implementation in the Mercurial plugin.  </p>

<p>It will be a different implementation with the git plugin, since git's concept of a reference repository is something that Mercurial does not have (as far as I know).</p>
</pre>
</details>

</ul>
</details>

<p>I believe Jesse Glick described that the SCMFileSystem implementation in the Mercurial plugin uses the initial checkout of the repository (for polling purposes) as a cache for other copies on the disc.  The concept was what was interesting to me, more than the specifics of the implementation in the Mercurial plugin.  </p>

<p>It will be a different implementation with the git plugin, since git's concept of a reference repository is something that Mercurial does not have (as far as I know).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=297569&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297569">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> Ah, I see. Thanks! Having an automatic cache for git repositories would be totally cool for Multi-branch pipelines. Last time I checked this project type (a long time ago), it created a new clone for each branch, which is not good for large repositories. Maybe it's been fixed somehow, dunno. Would be nice to have it as well.</p>

<p>This issue is unassigned right now. Do I have any chance to have it fixed soon?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> Ah, I see. Thanks! Having an automatic cache for git repositories would be totally cool for Multi-branch pipelines. Last time I checked this project type (a long time ago), it created a new clone for each branch, which is not good for large repositories. Maybe it's been fixed somehow, dunno. Would be nice to have it as well.</p>

<p>This issue is unassigned right now. Do I have any chance to have it fixed soon?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=297637&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297637">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I won't be working on this for a long time.  My focus is on evaluating and including the pull requests submitted by others, while tracking bug reports and resolving bugs.</p>
</pre>
</details>

</ul>
</details>

<p>I won't be working on this for a long time.  My focus is on evaluating and including the pull requests submitted by others, while tracking bug reports and resolving bugs.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=300823&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300823">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> I have an initial implementation of what you suggested. It's very short and simple. It works with certain limitations/assumptions (single SCM is used, all nodes are online, slave workspaces do not exist before the first run, etc).</p>

<p>I've extended GitSCM to use the same caches implemented in GitSCMSource for GitSCMFileSystem on master and manage similar caches on all slaves. Could you please have a look at it and advise me on improvements, so that we can have it in the upstream someday?</p>

<p>Surely I will remove the limitations. What bothers me is architectural decisions. E.g sharing caches with GitSCM might require refactoring the cache implementation out of AbstractGitSCMSource, which means some invasion into GitSCMFileSystem. It's not absolutely necessary because of the implementation details (cache entry is calculated from remote URL). But it might be a better implementation architecture wise. I'm willing to do it the best way.</p>

<p>The implementation: <a href="https://github.com/atikhono/git-plugin" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/atikhono/git-plugin</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> I have an initial implementation of what you suggested. It's very short and simple. It works with certain limitations/assumptions (single SCM is used, all nodes are online, slave workspaces do not exist before the first run, etc).</p>

<p>I've extended GitSCM to use the same caches implemented in GitSCMSource for GitSCMFileSystem on master and manage similar caches on all slaves. Could you please have a look at it and advise me on improvements, so that we can have it in the upstream someday?</p>

<p>Surely I will remove the limitations. What bothers me is architectural decisions. E.g sharing caches with GitSCM might require refactoring the cache implementation out of AbstractGitSCMSource, which means some invasion into GitSCMFileSystem. It's not absolutely necessary because of the implementation details (cache entry is calculated from remote URL). But it might be a better implementation architecture wise. I'm willing to do it the best way.</p>

<p>The implementation: <a href="https://github.com/atikhono/git-plugin" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/atikhono/git-plugin</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=300836&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300836">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=atikhono" class="user-hover" rel="atikhono">atikhono</a> thanks very much!  I will review it.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=atikhono" class="user-hover" rel="atikhono">atikhono</a> thanks very much!  I will review it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=302979&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-302979">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Actually no. The linked PR is implementing something quite unrelated: master-based caching. The stated bug here (really an RFE) is to expand environment variables in one Git extension.</p>
</pre>
</details>

</ul>
</details>

<p>Actually no. The linked PR is implementing something quite unrelated: master-based caching. The stated bug here (really an RFE) is to expand environment variables in one Git extension.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=302980&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-302980">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Now the root <em>desire</em> here may have been master-based caching. That is what <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> suggested as an alternative path to the solution you were working on. But the <em>stated issue</em> here is that a certain behavior does not expand variables, which would have been necessary for you to use your original workaround. So the JIRA situation here is all messed up.</p>
</pre>
</details>

</ul>
</details>

<p>Now the root <em>desire</em> here may have been master-based caching. That is what <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> suggested as an alternative path to the solution you were working on. But the <em>stated issue</em> here is that a certain behavior does not expand variables, which would have been necessary for you to use your original workaround. So the JIRA situation here is all messed up.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=304073&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304073">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/browse/JENKINS-44729" title="Optimize git clones via shared reference repo for multi branch builds of same repository" class="issue-link" data-issue-key="JENKINS-44729">JENKINS-44729</a> might better reflect what the posted PR is about.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/browse/JENKINS-44729" title="Optimize git clones via shared reference repo for multi branch builds of same repository" class="issue-link" data-issue-key="JENKINS-44729">JENKINS-44729</a> might better reflect what the posted PR is about.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=327476&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-327476">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've done a preliminary investigation of the original issue. Reference path expansion is done in hudson.plugins.git.extensions.impl.CloneOption.decorateCloneCommand:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">cmd.reference(build.getEnvironment(listener).expand(reference))
</pre>
</div></div>
<p>TaskListener doesn't get any other environment variables but:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">  BUILD_DISPLAY_NAME
  BUILD_ID
  BUILD_NUMBER
  BUILD_TAG
  BUILD_URL
  CLASSPATH
  HUDSON_HOME
  HUDSON_SERVER_COOKIE
  HUDSON_URL
  JENKINS_HOME
  JENKINS_SERVER_COOKIE
  JENKINS_URL
  JOB_BASE_NAME
  JOB_DISPLAY_URL
  JOB_NAME
  JOB_URL
  RUN_CHANGES_DISPLAY_URL
  RUN_DISPLAY_URL</pre>
</div></div>
<p>So if I reconfigure the job with JENKINS_HOME instead of HOME in 'Path of the reference repo to use during clone', it will expand that path properly.</p>
</pre>
</details>

</ul>
</details>

<p>I've done a preliminary investigation of the original issue. Reference path expansion is done in hudson.plugins.git.extensions.impl.CloneOption.decorateCloneCommand:</p>

<pre>
cmd.reference(build.getEnvironment(listener).expand(reference))
</pre>
<p>TaskListener doesn't get any other environment variables but:</p>

<pre>
  BUILD_DISPLAY_NAME
  BUILD_ID
  BUILD_NUMBER
  BUILD_TAG
  BUILD_URL
  CLASSPATH
  HUDSON_HOME
  HUDSON_SERVER_COOKIE
  HUDSON_URL
  JENKINS_HOME
  JENKINS_SERVER_COOKIE
  JENKINS_URL
  JOB_BASE_NAME
  JOB_DISPLAY_URL
  JOB_NAME
  JOB_URL
  RUN_CHANGES_DISPLAY_URL
  RUN_DISPLAY_URL</pre>
<p>So if I reconfigure the job with JENKINS_HOME instead of HOME in 'Path of the reference repo to use during clone', it will expand that path properly.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tario.jpg" title="tario's avatar" /> tario:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=332158&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332158">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>As none of those variables provide us with something that can be configure on a slave level, this is what we did as an (ugly) workaround. We define an environment variable called REFERENCE_REPO on each slave and the master.</p>

<p>Inside our jenkins pipeline library we create a step called checkoutCurrent</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-keyword">import</span> java.util.regex.Pattern
<span class="code-keyword">import</span> java.lang.reflect.Field
<span class="code-keyword">import</span> java.lang.reflect.Modifier

<span class="code-comment">// checkout current scm and resolve the REFERENCE_REPO environment variable, workaround <span class="code-keyword">for</span> JENKINS-43894
</span>def call(originalScm) {
	Field field = <span class="code-keyword">null</span>
	def updatedExtension = <span class="code-keyword">null</span>
	def originalReference = <span class="code-keyword">null</span>;

	<span class="code-keyword">try</span> {
		<span class="code-keyword">if</span> (originalScm.hasProperty(<span class="code-quote">'extensions'</span>) &amp;&amp; originalScm.extensions) {
			print(<span class="code-quote">'has extensions'</span>)
			def extensions = originalScm.extensions

			<span class="code-keyword">for</span> (<span class="code-object">int</span> i = 0; i &lt; extensions.size(); i++) {
				def extension = extensions[i]
				<span class="code-keyword">if</span> (extension.hasProperty(<span class="code-quote">'reference'</span>) &amp;&amp; extension.reference <span class="code-keyword">instanceof</span> <span class="code-object">String</span>) {
					updatedExtension = extension
					originalReference = extension.reference
					print(<span class="code-quote">'replacing reference: '</span> + originalReference)
					def reference = originalReference.replaceAll(Pattern.quote(<span class="code-quote">'${REFERENCE_REPO}'</span>), env.REFERENCE_REPO)
					print(<span class="code-quote">'with: '</span> + reference)

					<span class="code-comment">// https://gist.github.com/pditommaso/263721865d84dee6ebaf
</span>					field = extension.<span class="code-keyword">class.</span>getDeclaredField(<span class="code-quote">"reference"</span>)
					Field modifiersField = Field.<span class="code-keyword">class.</span>getDeclaredField(<span class="code-quote">"modifiers"</span>)
					modifiersField.setAccessible(<span class="code-keyword">true</span>)
					modifiersField.setInt(field, field.getModifiers() &amp; ~Modifier.FINAL)
					field.setAccessible(<span class="code-keyword">true</span>)
					field.set(extension, reference)
				}
			}
		}
		checkout changelog: <span class="code-keyword">false</span>, poll: <span class="code-keyword">false</span>, scm: originalScm
	} <span class="code-keyword">finally</span> {
		<span class="code-keyword">if</span> (field) {
			print(<span class="code-quote">'reset reference to: '</span> + originalReference)
			field.set(updatedExtension, originalReference)
		}
	}
}

</pre>
</div></div>

<p>and then in the pipeline instead of</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">checkout changelog: <span class="code-keyword">false</span>, poll: <span class="code-keyword">false</span>, scm: scm
</pre>
</div></div>

<p>we do</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">checkoutCurrent(scm)
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>As none of those variables provide us with something that can be configure on a slave level, this is what we did as an (ugly) workaround. We define an environment variable called REFERENCE_REPO on each slave and the master.</p>

<p>Inside our jenkins pipeline library we create a step called checkoutCurrent</p>


<pre>
<span class="code-keyword">import</span> java.util.regex.Pattern
<span class="code-keyword">import</span> java.lang.reflect.Field
<span class="code-keyword">import</span> java.lang.reflect.Modifier

<span class="code-comment">// checkout current scm and resolve the REFERENCE_REPO environment variable, workaround <span class="code-keyword">for</span> JENKINS-43894
</span>def call(originalScm) {
	Field field = <span class="code-keyword">null</span>
	def updatedExtension = <span class="code-keyword">null</span>
	def originalReference = <span class="code-keyword">null</span>;

	<span class="code-keyword">try</span> {
		<span class="code-keyword">if</span> (originalScm.hasProperty(<span class="code-quote">'extensions'</span>) && originalScm.extensions) {
			print(<span class="code-quote">'has extensions'</span>)
			def extensions = originalScm.extensions

			<span class="code-keyword">for</span> (<span class="code-object">int</span> i = 0; i < extensions.size(); i++) {
				def extension = extensions[i]
				<span class="code-keyword">if</span> (extension.hasProperty(<span class="code-quote">'reference'</span>) && extension.reference <span class="code-keyword">instanceof</span> <span class="code-object">String</span>) {
					updatedExtension = extension
					originalReference = extension.reference
					print(<span class="code-quote">'replacing reference: '</span> + originalReference)
					def reference = originalReference.replaceAll(Pattern.quote(<span class="code-quote">'${REFERENCE_REPO}'</span>), env.REFERENCE_REPO)
					print(<span class="code-quote">'with: '</span> + reference)

					<span class="code-comment">// https://gist.github.com/pditommaso/263721865d84dee6ebaf
</span>					field = extension.<span class="code-keyword">class.</span>getDeclaredField(<span class="code-quote">"reference"</span>)
					Field modifiersField = Field.<span class="code-keyword">class.</span>getDeclaredField(<span class="code-quote">"modifiers"</span>)
					modifiersField.setAccessible(<span class="code-keyword">true</span>)
					modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL)
					field.setAccessible(<span class="code-keyword">true</span>)
					field.set(extension, reference)
				}
			}
		}
		checkout changelog: <span class="code-keyword">false</span>, poll: <span class="code-keyword">false</span>, scm: originalScm
	} <span class="code-keyword">finally</span> {
		<span class="code-keyword">if</span> (field) {
			print(<span class="code-quote">'reset reference to: '</span> + originalReference)
			field.set(updatedExtension, originalReference)
		}
	}
}

</pre>

<p>and then in the pipeline instead of</p>

<pre>
checkout changelog: <span class="code-keyword">false</span>, poll: <span class="code-keyword">false</span>, scm: scm
</pre>

<p>we do</p>


<pre>
checkoutCurrent(scm)
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=332161&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332161">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tario" class="user-hover" rel="tario">tario</a> So 'scm' extensions can be overridden actually... awesome tip, thanks! I have submitted a PR to git-plugin some time ago: <a href="https://github.com/jenkinsci/git-plugin/pull/575." class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-plugin/pull/575.</a> No one who can merge seems to have time / be interested in it. I have just built the plugin with my patch and installed it in my instance.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tario" class="user-hover" rel="tario">tario</a> So 'scm' extensions can be overridden actually... awesome tip, thanks! I have submitted a PR to git-plugin some time ago: <a href="https://github.com/jenkinsci/git-plugin/pull/575." class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-plugin/pull/575.</a> No one who can merge seems to have time / be interested in it. I have just built the plugin with my patch and installed it in my instance.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tario.jpg" title="tario's avatar" /> tario:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=332195&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332195">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Updated my code a little bit, now it will revert the changes after doing the checkout (as otherwise this modified the general config I had in the job).</p>

<p>Would be really nice to get your PR merged.</p>
</pre>
</details>

</ul>
</details>

<p>Updated my code a little bit, now it will revert the changes after doing the checkout (as otherwise this modified the general config I had in the job).</p>

<p>Would be really nice to get your PR merged.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stevenfoster:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=334648&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-334648">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Does this only work on environment variables configured on the node in jenkins? and so wouldn't work for non-static agents i.e. launched from cloud provider?</p>

<p> </p>

<p>Or does it work on environment variables set on the agent itself?</p>
</pre>
</details>

</ul>
</details>

<p>Does this only work on environment variables configured on the node in jenkins? and so wouldn't work for non-static agents i.e. launched from cloud provider?</p>

<p> </p>

<p>Or does it work on environment variables set on the agent itself?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=334707&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-334707">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stevenfoster" class="user-hover" rel="stevenfoster">stevenfoster</a> For cloud slaves you will need to support setting environment variables via Node Properties in Jenkins. That should be implemented in a plugin that provides those slaves.</p>

<p>The patch above should work for environment variables set on an agent.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stevenfoster" class="user-hover" rel="stevenfoster">stevenfoster</a> For cloud slaves you will need to support setting environment variables via Node Properties in Jenkins. That should be implemented in a plugin that provides those slaves.</p>

<p>The patch above should work for environment variables set on an agent.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tario.jpg" title="tario's avatar" /> tario:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=352029&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-352029">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>As far as I can tell the fix for this <a href="https://github.com/jenkinsci/git-plugin/pull/575" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-plugin/pull/575</a> was merged to master, so should be in the next 4.0 based release.</p>
</pre>
</details>

</ul>
</details>

<p>As far as I can tell the fix for this <a href="https://github.com/jenkinsci/git-plugin/pull/575" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-plugin/pull/575</a> was merged to master, so should be in the next 4.0 based release.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=364163&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364163">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>By the way the use case described in the issue description seems like a bad idea to me. More straightforward to have the script determine the Git commit being built, and then rather than running <tt>checkout scm</tt> on each node, explicitly construct a <tt>GitSCM</tt> configured exactly the way you want, according to whatever Groovy logic you need. The kind of plugin-based environment variable expansion seen in this PR should only be necessary in freestyle projects.</p>
</pre>
</details>

</ul>
</details>

<p>By the way the use case described in the issue description seems like a bad idea to me. More straightforward to have the script determine the Git commit being built, and then rather than running <tt>checkout scm</tt> on each node, explicitly construct a <tt>GitSCM</tt> configured exactly the way you want, according to whatever Groovy logic you need. The kind of plugin-based environment variable expansion seen in this PR should only be necessary in freestyle projects.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tario.jpg" title="tario's avatar" /> tario:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=364172&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364172">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> maybe I'm missing something but to me this looks totally normal to use in pipeline jobs. 
We have some linux slaves and some windows slaves and they each have a different location of the reference repo. I can even think of cases where you have e.g. linux nodes which for whatever reason have different paths to the reference repo. Having this configurable on a slave level and then just use it in the multibranch config is perfect, that way the pipeline does not care about how the checkout happens.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> maybe I'm missing something but to me this looks totally normal to use in pipeline jobs. <br/>
We have some linux slaves and some windows slaves and they each have a different location of the reference repo. I can even think of cases where you have e.g. linux nodes which for whatever reason have different paths to the reference repo. Having this configurable on a slave level and then just use it in the multibranch config is perfect, that way the pipeline does not care about how the checkout happens.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hasselmm:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=364206&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364206">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Exactly that. Reference repo paths, library paths, tool paths: This all are node specific settings which have no place in the actual pipeline description. As an extreme<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/help_16.png" height="16" width="16" align="absmiddle" alt="" border="0"/> example imagine that your project is published on Github or the like. Do you really want to share with the world the gritty details of your local Jenkins configuration?</p>
</pre>
</details>

</ul>
</details>

<p>Exactly that. Reference repo paths, library paths, tool paths: This all are node specific settings which have no place in the actual pipeline description. As an extreme<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/help_16.png" height="16" width="16" align="absmiddle" alt="" border="0"/> example imagine that your project is published on Github or the like. Do you really want to share with the world the gritty details of your local Jenkins configuration?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=364225&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364225">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>library paths, tool paths</p></blockquote>

<p>These are of course better handled via Docker containers or the like if you can manage it.</p>

<blockquote><p>node specific settings which have no place in the actual pipeline description</p></blockquote>

<p>I did mean to imply that the <tt>Jenkinsfile</tt> must list actual filesystem paths on each possible node, merely that the program has <em>some way</em> of determining this information on the fly, preferably not involving Jenkins global configuration. That could be an environment variable set on the agent’s account, etc. The point is that the selection and configuration of a reference repo is something under the control of the job maintainer or Jenkins admin, rather than being <a href="https://github.com/jglick/jk--" class="external-link" target="_blank" rel="nofollow noopener">baked into a plugin</a>.</p>

<blockquote><p>imagine that your project is published on Github</p></blockquote>

<p>I do not have to imagine that, it is a daily reality. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<blockquote><p>Do you really want to share with the world the gritty details of your local Jenkins configuration?</p></blockquote>

<p>In the case of the Jenkins project, in fact <a href="https://github.com/jenkins-infra/documentation/blob/master/ci.adoc" class="external-link" target="_blank" rel="nofollow noopener">we do</a>, but in case you do not, there are <a href="https://jenkins.io/doc/book/pipeline/shared-libraries/" class="external-link" target="_blank" rel="nofollow noopener">various</a> <a href="https://github.com/jenkinsci/pipeline-multibranch-defaults-plugin/blob/1f2f0d7d33a0d1b058877dcc1357f1830dd2632c/README.md#how-it-works" class="external-link" target="_blank" rel="nofollow noopener">alternatives</a>.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>library paths, tool paths</p></blockquote>

<p>These are of course better handled via Docker containers or the like if you can manage it.</p>

<blockquote><p>node specific settings which have no place in the actual pipeline description</p></blockquote>

<p>I did mean to imply that the <tt>Jenkinsfile</tt> must list actual filesystem paths on each possible node, merely that the program has <em>some way</em> of determining this information on the fly, preferably not involving Jenkins global configuration. That could be an environment variable set on the agent’s account, etc. The point is that the selection and configuration of a reference repo is something under the control of the job maintainer or Jenkins admin, rather than being <a href="https://github.com/jglick/jk--" class="external-link" target="_blank" rel="nofollow noopener">baked into a plugin</a>.</p>

<blockquote><p>imagine that your project is published on Github</p></blockquote>

<p>I do not have to imagine that, it is a daily reality. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<blockquote><p>Do you really want to share with the world the gritty details of your local Jenkins configuration?</p></blockquote>

<p>In the case of the Jenkins project, in fact <a href="https://github.com/jenkins-infra/documentation/blob/master/ci.adoc" class="external-link" target="_blank" rel="nofollow noopener">we do</a>, but in case you do not, there are <a href="https://jenkins.io/doc/book/pipeline/shared-libraries/" class="external-link" target="_blank" rel="nofollow noopener">various</a> <a href="https://github.com/jenkinsci/pipeline-multibranch-defaults-plugin/blob/1f2f0d7d33a0d1b058877dcc1357f1830dd2632c/README.md#how-it-works" class="external-link" target="_blank" rel="nofollow noopener">alternatives</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>wgc123:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=383000&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383000">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Any updates on releasing this?  I verified it does not work on my v4.0 plugin (can resolve $JENKINS_HOME but not $HOME so is useless) and the changelog does <em>not</em> list it for that or v4.1 beta</p>

<p>We have pipelines across many nodes, including different platforms, so we can't take full advantage of reference repos (the workaround of manually re-defining the checkout for every stage would be maintenance disaster) — we need a way of dynamically building the reference repo path per node per pipeline.</p>
</pre>
</details>

</ul>
</details>

<p>Any updates on releasing this?  I verified it does not work on my v4.0 plugin (can resolve $JENKINS_HOME but not $HOME so is useless) and the changelog does <em>not</em> list it for that or v4.1 beta</p>

<p>We have pipelines across many nodes, including different platforms, so we can't take full advantage of reference repos (the workaround of manually re-defining the checkout for every stage would be maintenance disaster) — we need a way of dynamically building the reference repo path per node per pipeline.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=383002&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383002">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've proposed automatic repository caching as a <a href="https://docs.google.com/document/d/1M9Na1eJxwGZ7fY1L-PsBGv62jL_pPBtqmGePpwogUBw/edit#heading=h.pr5cab6hhyg" class="external-link" target="_blank" rel="nofollow noopener">Google Summer of Code idea</a> and have offered the pull request as a starting point for consideration.</p>

<p>Have you tried the alternative that is described by Jesse Glick? It seems like it would work in your case if you perform the environment variable expansion in the Jenkins checkout statement.</p>
</pre>
</details>

</ul>
</details>

<p>I've proposed automatic repository caching as a <a href="https://docs.google.com/document/d/1M9Na1eJxwGZ7fY1L-PsBGv62jL_pPBtqmGePpwogUBw/edit#heading=h.pr5cab6hhyg" class="external-link" target="_blank" rel="nofollow noopener">Google Summer of Code idea</a> and have offered the pull request as a starting point for consideration.</p>

<p>Have you tried the alternative that is described by Jesse Glick? It seems like it would work in your case if you perform the environment variable expansion in the Jenkins checkout statement.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tario.jpg" title="tario's avatar" /> tario:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=383005&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383005">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wgc123" class="user-hover" rel="wgc123">wgc123</a> did you actually specify HOME as an explicit variable on the node or are you just depending on an existing environment variable. If I remember correctly we're using this with some 4.* version for a while, but always with explicitly specified node variables.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wgc123" class="user-hover" rel="wgc123">wgc123</a> did you actually specify HOME as an explicit variable on the node or are you just depending on an existing environment variable. If I remember correctly we're using this with some 4.* version for a while, but always with explicitly specified node variables.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>wgc123:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-43894?focusedId=383008&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383008">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Good call <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tario" class="user-hover" rel="tario">tario</a>  It works!  I assumed I could just use environment variables that I believe existed, but whether it doesn't really exist, or the plugin only sees a subset of environment variables, it correctly found the reference repo using an env defined in the node!  It's not as clean as an automatically defined variable would be, but it looks like it works!  We have a good sized repo so even though we're on-prem, I was seeing 2+ minute fetches for those stages where the hard-coded path was wrong, but this brings it down to 15 seconds in my first few runs</p>

<p>Edit: forgot the most important part - this lets me define a single path (using forward slashes) that works on both Linux and Windows nodes.</p>
</pre>
</details>

</ul>
</details>

<p>Good call <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tario" class="user-hover" rel="tario">tario</a>  It works!  I assumed I could just use environment variables that I believe existed, but whether it doesn't really exist, or the plugin only sees a subset of environment variables, it correctly found the reference repo using an env defined in the node!  It's not as clean as an automatically defined variable would be, but it looks like it works!  We have a good sized repo so even though we're on-prem, I was seeing 2+ minute fetches for those stages where the hard-coded path was wrong, but this brings it down to 15 seconds in my first few runs</p>

<p>Edit: forgot the most important part - this lets me define a single path (using forward slashes) that works on both Linux and Windows nodes.</p>
